modules/ethereum-contract/builtin/src/lib.rs

```diff
@@ -0,0 +1,264 @@
+	BestSetDecode(codec::Error),
+	/// Failed to decode finality proof.
+	FinalityProofDecode(codec::Error),
+	/// Failed to verify justification.
+	JustificationVerify(ClientError),
+}
+
+/// Substrate header.
+#[derive(Debug)]
+pub struct Header {
+	/// Header hash.
+	pub hash: Hash,
+	/// Parent header hash.
+	pub parent_hash: Hash,
+	/// Header number.
+	pub number: BlockNumber,
+	/// GRANDPA validators change signal.
+	pub signal: Option<ValidatorsSetSignal>,
+}
+
+/// GRANDPA validators set change signal.
+#[derive(Debug)]
+pub struct ValidatorsSetSignal {
+	/// Signal delay.
+	pub delay: BlockNumber,
+	/// New validators set.
+	pub validators: Vec<u8>,
+}
+
+/// Convert from U256 to BlockNumber.
+pub fn to_substrate_block_number(number: U256) -> Result<BlockNumber, Error> {
+	match number == number.low_u32().into() {
+		true => Ok(number.low_u32()),
+		false => Err(Error::BlockNumberDecode),
+	}
+}
+
+/// Convert from BlockNumber to U256.
+pub fn from_substrate_block_number(number: BlockNumber) -> Result<U256, Error> {
+	Ok(U256::from(number as u64))
+}
+
+/// Parse Substrate header.
+pub fn parse_substrate_header(raw_header: &[u8]) -> Result<Header, Error> {
+	RuntimeHeader::decode(&mut &raw_header[..])
+		.map(|header| Header {
+			hash: header.hash(),
+			parent_hash: header.parent_hash,
+			number: header.number,
+			signal: sp_runtime::traits::Header::digest(&header)
+				.log(|log| {
+					log.as_consensus().and_then(|(engine_id, log)| {
+						if engine_id == GRANDPA_ENGINE_ID {
+							Some(log)
+						} else {
+							None
+						}
+					})
+				})
+				.and_then(|log| ConsensusLog::decode(&mut &log[..]).ok())
+				.and_then(|log| match log {
+					ConsensusLog::ScheduledChange(scheduled_change) => Some(ValidatorsSetSignal {
+						delay: scheduled_change.delay,
+						validators: scheduled_change.next_authorities.encode(),
+					}),
+					_ => None,
+				}),
+		})
+		.map_err(Error::HeaderDecode)
+}
+
+/// Verify GRANDPA finality proof.
+pub fn verify_substrate_finality_proof(
+	finality_target_number: BlockNumber,
+	finality_target_hash: Hash,
+	best_set_id: u64,
+	raw_best_set: &[u8],
+	raw_finality_proof: &[u8],
+) -> Result<(), Error> {
+	// TODO: accept other types of finality proof (Commit)
+	let best_set = AuthorityList::decode(&mut &raw_best_set[..]).map_err(Error::BestSetDecode)?;
+
+	substrate::GrandpaJustification::decode_and_verify_finalizes(
+		&raw_finality_proof,
+		(finality_target_hash, finality_target_number),
+		best_set_id,
+		&best_set.into_iter().collect(),
+	)
+	.map_err(Error::JustificationVerify)
+	.map(|_| ())
+}
+
+/// ==================================================================================================================
+/// === Everything below should be (?) exposed by Substrate ==========================================================
+/// ==================================================================================================================
+
+mod substrate {
```

Comment:
@HCastano @tomusdrw This is the most important question about this PR - how's better to organize this check? This code is now well isolated in single [Substrate file](https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/justification.rs), but all methods are private. I think it may be helpful for other developers as well.

IMO it is tightly linked to what you're working on, @HCastano ? I mean that finality subscription RPC - what's going on with justifications that are read using this subcription? Are they blindly trusted, or verified somewhere else? Just want to synchronize here, so we won't make similar changes to Substrate code :)

For using in builtin, we have 2 options. The worst is to duplicate code here (this is what you're seeing now) - it depends on Substrate internals => let's not use it. So the only option (imo) is to make `GrandpaJustification::decode_and_verify()` public (probably extracting this whole struct in separate crate before). Wdyt?

Reply:
https://github.com/paritytech/substrate/pull/5746